node/src/signing/recent_blocks_tracker.rs

```diff
@@ -0,0 +1,943 @@
+    /// we ensure that the final head is not pruned. Otherwise, not only would the logic be very
+    /// messy, but also we would not be able to provide a contiguous stream of finalized blocks.
+    fn minimum_height_to_keep(&self) -> Option<u64> {
+        let Some(final_head) = &self.final_head else {
+            return None;
+        };
+        Some(
+            self.maximum_height_available
+                .saturating_sub(self.window_size)
+                .add(1)
+                .min(final_head.height),
+        )
+    }
+
+    /// Remove old blocks that are neither needed for the `classify_block` query or for providing
+    /// the stream of finalized blocks.
+    fn prune_old_blocks(&mut self) {
+        let Some(minimum_height_to_keep) = self.minimum_height_to_keep() else {
+            return;
+        };
+        // Note: unwrap_or cannot fail because if we have a minimum height then we have at least one
+        // block. Still, we'll program defensively.
+        if self
+            .root_children
+            .iter()
+            .map(|child| child.height)
+            .min()
+            .unwrap_or(0)
+            >= minimum_height_to_keep
+        {
+            return;
+        }
+
+        let mut new_root_children = Vec::new();
+        let mut old_nodes = Vec::new();
+        for child in &self.root_children {
+            child.closest_descendants_with_height_at_least(
+                minimum_height_to_keep,
+                &mut new_root_children,
+                &mut old_nodes,
+            );
+        }
+        for child in &new_root_children {
+            *child.parent.lock().unwrap() = None;
+        }
+        self.root_children = new_root_children;
+        for old_node in old_nodes {
+            self.hash_to_node.remove(&old_node.hash);
+            self.node_to_content.remove(&old_node.hash);
+        }
+    }
+
+    /// Classifies a block into one of the categories in `CheckBlockResult`.
+    pub fn classify_block(&self, block_hash: CryptoHash, block_height: u64) -> CheckBlockResult {
+        if self
+            .maximum_height_available
+            .saturating_sub(self.window_size)
+            .add(1)
+            > block_height
+        {
+            return CheckBlockResult::OlderThanRecentWindow;
+        }
+        match self.hash_to_node.get(&block_hash) {
+            Some(node) => {
+                if node.is_final.load(Ordering::Relaxed) {
+                    return CheckBlockResult::RecentAndFinal;
+                }
+                if let Some(final_head) = &self.final_head {
+                    if block_height <= final_head.height {
+                        return CheckBlockResult::NotIncluded;
+                    }
+                }
+                if node.canonical.load(Ordering::Relaxed) {
+                    return CheckBlockResult::OptimisticAndCanonical;
+                }
+                CheckBlockResult::OptimisticButNotCanonical
```

Comment:
For a block that is not final, why do we need to distinguish between it being on the canonical chain and it not being on the canonical chain?

Reply:
Signature requests on the canonical chain are attempted, but those not on the canonical chain are not. This is to handle the case where the blockchain may be struggling and having forks, and we don't want to doubly or triply or quadruply etc. attempting the same signature requests because they exist on multiple forks.

Reply:
I see. Makes sense